jans-cedarling/cedarling/src/common/policy_store.rs

```diff
@@ -5,155 +5,430 @@
  */
 
-use std::collections::HashMap;
-
 use super::cedar_schema::CedarSchema;
+use base64::prelude::*;
+use cedar_policy::PolicyId;
+use serde::{de::Visitor, Deserialize, Deserializer};
+use std::{collections::HashMap, fmt};
 
-/// PolicyStoreMap it is a store for `PolicyStore` accessible by key.
-#[derive(Debug, serde::Deserialize)]
-pub(crate) struct PolicyStoreMap {
-    #[serde(flatten)]
-    pub policy_stores: HashMap<String, PolicyStore>,
-}
-
-/// PolicyStore contains all the data the Cedarling needs to verify JWT tokens and evaluate policies
+/// Represents the store of policies used for JWT validation and policy evaluation in Cedarling.
+///
+/// The `PolicyStore` contains the schema and a set of policies encoded in base64,
+/// which are parsed during deserialization.
 #[derive(Debug, Clone, serde::Deserialize)]
 pub struct PolicyStore {
-    pub schema: CedarSchema,
-    #[serde(deserialize_with = "deserialize_policy_set::parse")]
-    pub policies: cedar_policy::PolicySet,
+    /// The cedar version to use when parsing the schema and policies.
+    #[serde(deserialize_with = "check_cedar_version")]
+    #[allow(dead_code)]
+    pub cedar_version: String,
+
+    /// Cedar schema in base64-encoded string format.
+    ///
+    /// Extracted from the `policy_store.json` file.
+    pub cedar_schema: CedarSchema,
+
+    /// Cedar policy set in base64-encoded string format.
+    ///
+    /// Extracted from the `policy_store.json` file and deserialized using `deserialize_policies`.
+    #[serde(deserialize_with = "parse_cedar_policy")]
+    pub cedar_policies: cedar_policy::PolicySet,
+
+    /// An optional list of trusted issuers.
+    ///
+    /// This field may contain issuers that are trusted to provide tokens, allowing for additional
+    /// verification and security when handling JWTs.
+    #[allow(dead_code)]
+    pub trusted_issuers: Option<Vec<TrustedIssuer>>,
 }
 
-// Deserialization to [`cedar_policy::PolicySet`] moved here
-// to be close to structure where it is used.
-// And current module is small so it is OK to store it here.
-mod deserialize_policy_set {
+/// Represents a trusted issuer that can provide JWTs.
+///
+/// This struct includes the issuer's name, description, and the OpenID configuration endpoint
+/// for discovering issuer-related information.
+#[derive(Debug, Clone, Deserialize)]
+#[allow(dead_code)]
+pub struct TrustedIssuer {
+    /// The name of the trusted issuer.
+    pub name: String,
+
+    /// A brief description of the trusted issuer.
+    pub description: String,
+
+    /// The OpenID configuration endpoint for the issuer.
+    ///
+    /// This endpoint is used to obtain information about the issuer's capabilities.
+    pub openid_configuration_endpoint: String,
+
+    /// Optional metadata related to the tokens issued by this issuer.
+    ///
+    /// This field may include role mappings that help define the access levels for issued tokens.
+    #[serde(deserialize_with = "check_token_metadata")]
```

Comment:
Why we can't implement `serde::Deserialize` trait for `TokenMetadata` ?

Reply:
i can imlement `Deserialize` for `TokenMetadata` but i can't implement Deserialize for `Option` or `Vec`. Here's the line you omitted from your comment:
```rust
    /// Optional metadata related to the tokens issued by this issuer.
    ///
    /// This field may include role mappings that help define the access levels for issued tokens.
    #[serde(deserialize_with = "check_token_metadata")]
    pub token_metadata: Option<Vec<TokenMetadata>>,
```
Here's what I tried to do at first:
```rust
/// This implementation ensures that the metadata contains at most one `TokenMetadata` with a `role_mapping`
/// to prevent ambiguous role assignments.
impl<'de> Deserialize<'de> for Vec<TokenMetadata> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let token_metadata: Vec<TokenMetadata> = Vec::deserialize(deserializer)?;

        let count = token_metadata
            .iter()
            .filter(|token| token.role_mapping.is_some())
            .count();

        if count > 1 {
            return Err(serde::de::Error::custom(
                "there can only be one TokenMetadata with a role_mapping".to_string(),
            ));
        }

        Ok(token_metadata)
    }
}
```
you will get a
> `Vec` is not defined in the current crate

from this

Reply:
basically, i am also doing a check over the deserialization of the whole `Vec` since we should only have one Metadata with the `role_mapping` field.

Reply:
OK, it make sence